odl/space/fspace.py

```diff
@@ -154,196 +356,346 @@ class FunctionSetVector(Operator):
-                raise TypeError('input is neither an element of the '
-                                'function domain {!r} nor an array or '
-                                'meshgrid-type coordinate list.'
-                                ''.format(self.domain))
-
-            if (min_coords not in self.domain or
-                    max_coords not in self.domain):
-                raise ValueError('input contains points outside '
-                                 '`domain` {}.'.format(self.domain))
-
-        out = self._call(*x)
+        out : range element or array of elements
+            Result of the function evaluation
 
-        if not (out in self.range or
-                (isinstance(out, np.ndarray) and
-                    out.flat[0] in self.range)):
-            raise TypeError('result {!r} not an element or an array of '
-                            'elements of the function range {}.'
-                            ''.format(out, self.range))
+        Raises
+        ------
+        TypeError
+            If `x` is not a valid vectorized evaluation argument
 
-        return out
+            If `out` is not a range element or a `numpy.ndarray`
+            of range elements
 
-    def apply(self, out, *x):
-        """Vectorized and multi-argument in-place evaluation.
+        ValueError
+            If evaluation points fall outside the valid domain
+        """
+        vec_bounds_check = kwargs.pop('vec_bounds_check', self.vectorized)
+        if vec_bounds_check and not hasattr(self.domain, 'contains_all'):
+            raise AttributeError('vectorized bounds check not possible for '
+                                 'domain {}, missing `contains_all()` '
+                                 'method.'.format(self.domain))
+        scalar_out = False
+
+        # A. Pre-checks and preparations
+        # 1. vectorized? (a/b)
+        # 2a. x = domain element (1), array (2), meshgrid (3)
+        # 2a1. make x a (d, 1) array; set a flag that output shall be
+        #      scalar; apply case 2a2
+        # 2a2. out_shape = (x.shape[1],)
+        # 2a3. out_shape = (x[0].shape[1],) if ndim == 1 else
+        #      np.broadcast(*x).shape
+        # 2a. (cont.) If vec_bounds_check, check domain.contains_all(x)
+        # 2b. x in domain? -> yes ok, no error; out is None? yes -> ok,
+        #     no -> error
+        #
+        # B. Evaluation and post-checks
+        # 1. out is None? (a/b)
+        # 1a. out = call(x)
+        #     vectorized? (1/2)
+        # 1a1. out.shape == out_shape? -> error if no
+        #      If vec_bounds_check, check range.contains_all(out)
+        # 1a2. nothing
+        # 2b. vectorized? (1/2)
+        # 2b1. out is array and out.shape == out_shape? -> error if no;
+        #     call(x, out=out);
+        #     If vec_bounds_check, check range.contains_all(out)
+        # 2b2. error (out given but not vectorized)
+
+        # TODO: simplify logic in second part
+        if self.vectorized:
```

Comment:
Yes, that makes sense. If we decide to always vectorize, this will become significantly simpler, though.